### PR TITLE
streamline textual output of X509 certs and CRLs: remove empty lines, add/remove tag, correct indentation

### PR DESCRIPTION
**Pull request by DDvO**
This makes X.509 printing slightly more consistent.

When viewing textual certificate output, e.g., of the `x509` app, I was often wondering why there is an empty line between the output of "X509v3 Authority Key Identifier:" and "X509v3 Basic Constraints:".
Moreover, the value printed for "X509v3 Authority Key Identifier:" is prepended by the tag `keyid`", which is not the case for "X509v3 Subject Key Identifier:", such that the output in particular for self-signed certs looks a bit odd, for instance:
```
            X509v3 Subject Key Identifier:
                11:49:46:19:2A:4E:4D:D1:C8:FB:79:55:3D:81:99:22:EE:34:4F:22
            X509v3 Authority Key Identifier:
                keyid:11:49:46:19:2A:4E:4D:D1:C8:FB:79:55:3D:81:99:22:EE:34:4F:22

            X509v3 Basic Constraints:
                CA:TRUE
```
This PR streamlines such output to, in the above example,
```
            X509v3 Subject Key Identifier:
                11:49:46:19:2A:4E:4D:D1:C8:FB:79:55:3D:81:99:22:EE:34:4F:22
            X509v3 Authority Key Identifier:
                11:49:46:19:2A:4E:4D:D1:C8:FB:79:55:3D:81:99:22:EE:34:4F:22
            X509v3 Basic Constraints:
                CA:TRUE
```
The tag `keyid:` is left out only if the X509v3 Authority Key Identifier, as usual, does not also include an `issuer` or `serial` sub-field.
Also other (less common) occurrences of empty lines are eliminated. This was more effort than expected because the handling of newlines is implicit and spread across many places. I did not recognize a consistent design behind it.

I also found that the output of signature values at the end of certs and CRLs is slightly inconsistent because there no tag is given for the `signatureValue` and the indentation of the actual value is 9 rather than 8 (= 2 * 4). This is also corrected, such that, for instance,
```
    Signature Algorithm: sha256WithRSAEncryption
         04:8f:c3:77:48:06:29:c0:8d:66:2e:6b:48:a3:b3:e0:dd:5b:
         0a:e7:a4:0b:7e:72:91:fc:37:29:7f:81:1e:60:66:7b:ba:94:
```
becomes
```
    Signature Algorithm: sha256WithRSAEncryption
    Signature Value:
        04:8f:c3:77:48:06:29:c0:8d:66:2e:6b:48:a3:b3:e0:dd:5b:
        0a:e7:a4:0b:7e:72:91:fc:37:29:7f:81:1e:60:66:7b:ba:94:
```
The overall effect of this PR on printing certs can be seen, at the example of one of the test certs at https://github.com/openssl/openssl/compare/master...siemens:X509V3_EXT_val_prn-multiline-fix?expand=1#diff-da7c216b91096b818848e5b53012606c

- [x] tests are added or updated

### CODE REVIEW

File: crypto/x509v3/v3_akey.c [hunk 1]
```diff
@@ -42,7 +42,7 @@ static STACK_OF(CONF_VALUE) *i2v_AUTHORITY_KEYID(X509V3_EXT_METHOD *method,
     char *tmp;
     if (akeyid->keyid) {
         tmp = OPENSSL_buf2hexstr(akeyid->keyid->data, akeyid->keyid->length);
-        X509V3_add_value("keyid", tmp, &extlist);
+        X509V3_add_value(akeyid->issuer || akeyid->serial ? "keyid" : NULL, tmp, &extlist);
```
Comment (levitte) [inline crypto/x509v3/v3_akey.c]: What does `X509V3_add_value` do when the first arg is NULL?

Comment (DDvO) [reply crypto/x509v3/v3_akey.c]: Pleased that you had a look at this PR.

Would be good if functions like `X509V3_add_value`, which appear in the public headers, were documented ;-)

I has been a while that I've been doing this patch, but I'm pretty sure that a `NULL` pointer is fine at this place, and then simply no name is printed, for several reasons:
* This function does get called with `NULL` `name` arg also elsewhere, for instance:
```
crypto/x509v3/v3_tlsf.c:            X509V3_add_value(NULL, tls_feature_tbl[j].name, &ext_list);
crypto/x509v3/v3_tlsf.c:            X509V3_add_value_int(NULL, ai, &ext_list);
```
* The function checks the `name` parameter for `NULL` in way that suggests that this parameter is optional.
* As far as I recall I tried out also the `NULL` case.

Comment (DDvO) [reply crypto/x509v3/v3_akey.c]: BTW, the idea for calling `X509V3_add_value()` at the above quoted place with an `NULL` pointer in case neither `keyid->issuer` nor `akeyid->serial` are present was that in this case it should be clear that the hex value printed is just a simple authority key identifier, and then the value can be printed in the very same way as for the subject key identifier. In the given example, 
```
            X509v3 Subject Key Identifier:
                11:49:46:19:2A:4E:4D:D1:C8:FB:79:55:3D:81:99:22:EE:34:4F:22
            X509v3 Authority Key Identifier:
                keyid:11:49:46:19:2A:4E:4D:D1:C8:FB:79:55:3D:81:99:22:EE:34:4F:22
```
becomes simpler (and to most user less confusing) to read:
```
            X509v3 Subject Key Identifier:
                11:49:46:19:2A:4E:4D:D1:C8:FB:79:55:3D:81:99:22:EE:34:4F:22
            X509v3 Authority Key Identifier:
                11:49:46:19:2A:4E:4D:D1:C8:FB:79:55:3D:81:99:22:EE:34:4F:22
```

File: crypto/x509/t_x509.c [hunk 2]
```diff
@@ -302,25 +302,28 @@ int X509_signature_print(BIO *bp, const X509_ALGOR *sigalg,
                          const ASN1_STRING *sig)
 {
     int sig_nid;
-    if (BIO_puts(bp, "    Signature Algorithm: ") <= 0)
+    int indent = 4;
```
Comment (mattcaswell) [inline crypto/x509/t_x509.c]: Perhaps this should be const? In any case please add a blank line between the declarations and the rest of the code.

File: crypto/x509/t_x509.c [hunk 3]
```diff
@@ -302,25 +302,28 @@ int X509_signature_print(BIO *bp, const X509_ALGOR *sigalg,
...
             ameth = EVP_PKEY_asn1_find(NULL, pkey_nid);
             if (ameth && ameth->sig_print)
-                return ameth->sig_print(bp, sigalg, sig, 9, 0);
+                return ameth->sig_print(bp, sigalg, sig, indent + 4, 0);
         }
     }
-    if (sig)
-        return X509_signature_dump(bp, sig, 9);
-    else if (BIO_puts(bp, "\n") <= 0)
+    if (BIO_write(bp, "\n", 1) != 1)
```
Comment (mattcaswell) [inline crypto/x509/t_x509.c]: This looks odd given you've used BIO_puts for the same thing in other places.

File: crypto/x509v3/v3_akey.c (hunk 1 above)
Comment (mattcaswell) [reply crypto/x509v3/v3_akey.c]: Could we add some parens in here (prior to ?).
I actually find it a bit weird that we would sometimes print keyid and not other times. Would it be better for them both to always print keyid, or (alternatively) for neither of them to ever do it.

Comment (DDvO) [reply crypto/x509v3/v3_akey.c]: Sure we could place parentheses around `keyid->issuer || akeyid->serial` 
but according to your comment I meanwhile also see that for uniformity the `keyid:` tag should better be generally left out (which for instance is done also by Mozilla's cert viewer). 
Since moreover the interpretation of the raw ID data should be obvious from the line preceding it (`X509v3 Authority Key Identifier:`) I've just removed the `keyid:` tag unconditionally.

File: crypto/x509/t_x509.c (hunk 2 above)
Comment (DDvO) [reply crypto/x509/t_x509.c]: Yes - done.

**Comment by DDvO** [comment]
I had overlooked handling one minor change request, which I just have done.

I've also rebased this PR (the CI builds went through except for two cases with some unrelated package issue with Travis CI).

@mattcaswell, is there anything more to do?

**Comment by mattcaswell** [comment]
Travis failures are not relevant. Needs second review.

**Comment by DDvO** [comment]
Thanks @levitte for the second review&approval.
I've meanwhile rebased the PR again and the (unrelated) Travis CI issue is gone.
Yet this time AppVeyor got stuck (while it worked fine last time).

**Comment by mattcaswell** [comment]
Appveyor failure is unrelated to this PR though.

**Comment by mattcaswell** [comment]
Pushed - thanks. I squashed the last commit.

**Comment by mlindner** [comment]
This fixed https://github.com/openssl/openssl/issues/7284.

**Comment by darrenmoffat** [comment]
I've just discovered this change as a result of changing the Oracle ZFS Storage Appliance code that manages certificates over from 1.0.2 to 3.0.10. The area of code that was impacted for us is turning the certificate into our XMLRPC/JSON format for display in the UI, for the authorityKeyIdentity. We generate an array of name/value pairs because you need to know what the value represents, eg keyid, issuer, serial etc.

It seems that the original intent was to change the output of the openssl(1) command and rightly so there was concern raised about breaking applications.  I agree that nothing should be attempting to parse the human readable output of openssl(1). As such I would have expected this change should have been implemented in openssl(1), instead of deep inside the x509 code in libcrypto.

The way this change has been implemented it actually breaks the API callers.  While I understand and appreciated there was no expectation of API/ABI stability between 1.0.x and 3.0.x this change is very subtle and was very difficult to track down.
It seems wrong to me that the API was changed to no longer report that the name of value is 'keyid' and just return the value with a name of NULL if it is the only authorityKeyIdentity present.  So unless I'm missing something, what happens now is that the 'keyid' valuename is only added if either or both of issuer/serial are also present.  So the caller now needs to determine if the valuename was NULL and know that means the value is a 'keyid' otherwise look at the name to determine if the value is a serial or issuer based on the non-NULL valuename.  Is that correct ?

**Comment by t8m** [comment]
Yes, it is correct.

Unfortunately it would be basically impossible to make the change in openssl(1) only. On the other hand it is true that the conditional removal of the `keyid:` prefix makes this really hard to parse.

**Comment by DDvO** [comment]
Sorry to hear that this change of 4+ years ago caused some trouble.
In retrospect, it should have been mentioned in `CHANGES.md`
(though I'm not sure if you had a look at it when tracking down the issue).

Correct that the change affects the libcrypto API, not only the apps. Yet as I wrote in https://github.com/openssl/openssl/pull/6226#issuecomment-458246971
also the output of those API functions is meant for human consumption, not for machine parsing.

Unfortunately, `X509_print()` and friends are still undocumented,
but from their names it should be clear that one better should not base a cert (format) conversion function on them.
